helix-core/src/main/java/org/apache/helix/manager/zk/ZkBaseDataAccessor.java

```diff
@@ -142,22 +142,22 @@ public ZkBaseDataAccessor(RealmAwareZkClient zkClient) {
   }
 
   private ZkBaseDataAccessor(Builder<T> builder) {
-    switch (builder.realmMode) {
+    switch (builder.getRealmMode()) {
       case MULTI_REALM:
         try {
-          if (builder.zkClientType == ZkClientType.DEDICATED) {
+          if (builder._zkClientType == ZkClientType.DEDICATED) {
```

Comment:
Following the previous comment, I can see this zkClientType is an exception which is not common for all the GenericZkHelixApiBuilder child classes' logic. In this case, we can let the ZkBaseDataAccessor builder extends my proposed GenericZkHelixApiBuilder method which creates a ZkClient, and in this extended method, we read and apply this _zkClientType.

Reply:
Or, if there are more than one builders require zkClientType, we can have another generic builder which includes the zkClientType logic and let it to extend GenericZkHelixApiBuilder.

Reply:
So as for the logic that creates which ZkClient to use -I've actually considered an option to include it in the builder, but I think since this is a construction logic for the API, I've explicitly decided to leave it in the constructor

My philosophy is that construction logic ("business logic" for the data accessors) should be left in the constructor, and builder should only set the parameters and validate.

But I see some duplicate logic in ZkBaseDataAccessor and ZkCacheBaseDataAccessor -

So I can actually abstract that logic out into ZkBaseDataAccessor.createZkClient(), and make it a static method that can be used in both accessors. That way, 1) we still keep the construction logic in the accessors 2) Builders have a clean separation and only contain parameter setters and validation logic, and 3) we reduce code duplication.

Reply:
Just to summarize:

This is the construction logic that only pertains to ZkBaseDataAccessor/ZkCacheBaseDataAccessor

Your feedback was that this is duplicated in both ZkBaseDataAccessor and ZkCacheBaseDataAccessor

I agree with your observation

Now we have a few options in terms of how to refactor this logic

Make it a reusable method that pertains to data accessors

2. Put this in the Builder

Here, between 1 and 2, my preference is 1, because this seems like business logic that belongs to the data accessor, not the builder.

What do you think?

Reply:
I would prefer 2. Once a class is designed with a builder, the builder should contain most of the complicated building logic. IMHO, including which client to use. Then the real private constructor of the Accessor just takes whatever client it is given. Although both ways avoid duplicate code, method 2 encapsulates the logic more closely.
Actually, it is an option that you put this logic into the generic builder. For the other accessors, this type can be fixed in the implemented builder class.

Reply:
I value your opinion and it also makes sense. For now, I can move all construction logic to the builders. I have updated the PR accordingly.